File: src/test/java/io/mysocialapp/client/java/ConversationTest.java
package io.mysocialapp.client.java;

import io.mysocialapp.client.AbstractNotificationCallback;
import io.mysocialapp.client.MySocialApp;
import io.mysocialapp.client.Session;
import io.mysocialapp.client.models.Conversation;
import io.mysocialapp.client.models.ConversationMessage;
import io.mysocialapp.client.models.ConversationMessagePost;
import io.mysocialapp.client.models.User;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.util.Set;
import java.util.stream.StreamSupport;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Created by evoxmusic on 31/07/2018.
 */
public class ConversationTest {

    private final static String APP_ID = "u470584465854a728453";
    private final static String EMAIL = "alice.jeith@mysocialapp.io";
    private final static String PASSWORD = "myverysecretpassw0rd";

    private final static String JOHN_EMAIL = "john.paul@mysocialapp.io";

    private final static Session session = new MySocialApp.Builder()
            .setAppId(APP_ID)
            .build()
            .blockingConnect(EMAIL, PASSWORD);

    private final static Session session2 = new MySocialApp.Builder()
            .setAppId(APP_ID)
            .build()
            .blockingConnect(JOHN_EMAIL, PASSWORD);

    private File getFile(String filePath) {
        return new File(System.class.getResource(filePath).getFile());
    }

    private Conversation getFirstConversation() {
        return StreamSupport.stream(session.getConversation().blockingStream(1).spliterator(), false).findFirst().get();
    }

    @Test
    void listConversations() {
        Iterable<Conversation> conversations = session.getConversation().blockingList(0, 10);
        assertNotNull(conversations);
    }

    @Test
    void createConversation() {
        User user = session.getUser().blockingStream(1).iterator().next().getUsers().get(0);

        Conversation conversation = new Conversation.Builder()
                .setName("SDK conversation")
                .addMember(user)
                .build();

        Conversation createdConversation = session.getConversation().blockingCreate(conversation);
        assertNotNull(createdConversation);
    }

    @Test
    void updateConversation() {
        User user = session.getUser().blockingStream(1).iterator().next().getUsers().get(0);

        Conversation conversation = new Conversation.Builder()
                .setName("SDK conversation")
                .addMember(user)
                .build();

        Conversation createdConversation = session.getConversation().blockingCreate(conversation);

        String cName = "Changed conversation name";
        createdConversation.setName(cName);
        createdConversation.blockingSave();

        assertEquals(session.getConversation().blockingGet(createdConversation.getId()).getName(), cName);
    }

    @Test
    void listMembersFromConversation() {
        Conversation conversation = getFirstConversation();
        Set<User> members = conversation.getMembers();
        assertTrue(members.size() > 0);
    }

    @Test
    void addMember() {

    }

    @Test
    void kickMember() {

    }

    @Test
    void listMessages() {
        Iterable<ConversationMessage> messages = getFirstConversation().getMessages().blockingStream(100);
        for (ConversationMessage conversationMessage : messages) {
            System.out.println(conversationMessage.getMessage());
        }
    }

    @Test
    void postMessage() {
        ConversationMessagePost conversationMessagePost = new ConversationMessagePost.Builder()
                .setMessage("This is a conversation message from SDK #top")
                .build();

        getFirstConversation().blockingSendMessage(conversationMessagePost);
    }

    @Test
    void postMessage_with_image() {
        ConversationMessagePost conversationMessagePost = new ConversationMessagePost.Builder()
                .setMessage("This is a conversation message with Image from SDK #yeah")
                .setImage(getFile("/hello.jpg"))
                .build();

        getFirstConversation().blockingSendMessage(conversationMessagePost);
    }

    @Test
    void quitConversation() {
        getFirstConversation().blockingQuit();
    }

    @Test
    void getUnreadMessages() {
        getFirstConversation().getMessages().getTotalUnreads();
    }

    @Test
    void handleConversationMessage() throws InterruptedException {
        session.getNotification().addNotificationListener(new AbstractNotificationCallback() {
            @Override
            public void onConversationMessage(@NotNull ConversationMessage conversationMessage) {
                System.out.println("received message: \"" + conversationMessage.getMessage()
                        + "\" from " + conversationMessage.getOwner().getDisplayedName());
            }
        });

        // message sent from another user
        User user = session2.getUser().blockingGet(session.getAccount().blockingGet().getId());

        ConversationMessagePost cmp = new ConversationMessagePost.Builder()
                .setMessage("hey [[user:" + user.getId() + "]]")
                .build();

        user.blockingSendPrivateMessage(cmp);

        Thread.sleep(10000L);
    }

    @Test
    void getTotalUnreadConversations() {
        assertNotNull(session.getConversation().blockingTotalUnread());
    }

}

File: src/test/kotlin/io/mysocialapp/client/ConversationTest.kt
package io.mysocialapp.client

import io.mysocialapp.client.models.Conversation
import io.mysocialapp.client.models.ConversationMessagePost
import org.junit.Test

/**
 * Created by evoxmusic on 24/05/2018.
 */
class ConversationTest {

    companion object {
        const val APP_ID = "u470584465854a194805"
    }

    private fun getSession(): Session? = MySocialApp.Builder().setAppId(APP_ID).build().connect("AliceX", "myverysecretpassw0rd")

    @Test
    fun `create conversation with 3 people`() {
        val s = getSession()

        val people = s?.user?.blockingStream(3)?.toList()?.get(0)?.users?.toSet() ?: emptySet()

        val conversation = Conversation.Builder()
                .setName("talk about the next event")
                .addMembers(people)
                .build()

        val createdConversation = s?.conversation?.blockingCreate(conversation)
        assert(createdConversation != null)
    }

    @Test
    fun `kick one person from this conversation`() {
        val s = getSession()
        val acc = s?.account?.blockingGet()
        val conversion = s?.conversation?.blockingList()?.firstOrNull()
        val kickedMember = conversion?.members?.find { it.id != acc?.id }?.let { conversion.blockingKickMember(it) }
        assert(kickedMember != null)
    }

    @Test
    fun `invite someone to this conversation`() {
        val s = getSession()
        val acc = s?.account?.blockingGet()

        // take 100 users, shuffle them, take the first that is not myself
        val userToInvite = s?.user?.blockingStream(100)?.map { it.users }?.flatMap {
            it?.asIterable() ?: emptyList()
        }?.shuffled()?.find { it.id != acc?.id }


        val conversion = s?.conversation?.blockingList()?.firstOrNull()

        val addedMember = userToInvite?.let { conversion?.blockingAddMember(it) }
        assert(addedMember != null)
    }

    @Test
    fun `rename the conversation`() {
        val s = getSession()
        val conversion = s?.conversation?.blockingList()?.firstOrNull()
        conversion?.name = "new conversation title :)"
        conversion?.blockingSave()

        assert(s?.conversation?.blockingGet(conversion?.id ?: -1L)?.name == "new conversation title :)")
    }

    @Test
    fun `get my conversations`() {
        val s = getSession()
        val conversations = s?.conversation?.blockingList()?.toList()
        assert(conversations != null)
    }

    @Test
    fun `get conversation messages`() {
        val s = getSession()
        val conversationMessages = s?.conversation?.blockingList()?.firstOrNull()?.messages?.blockingStream(33)?.toList()
        assert(conversationMessages != null)
    }

    @Test
    fun `get unread conversation messages`() {
        val s = getSession()
        val conversationMessages = s?.conversation?.blockingList()?.firstOrNull()?.messages?.blockingStreamAndConsume(33)?.toList()
        assert(conversationMessages != null)
    }

    @Test
    fun `add message to the last conversation`() {
        val s = getSession()
        val lastConversation = s?.conversation?.blockingList()?.firstOrNull()

        val message = ConversationMessagePost.Builder()
                .setMessage("Hello, this is a message from our SDK")
                .build()

        val messageSent = lastConversation?.blockingSendMessage(message)
        assert(messageSent != null)
    }

    @Test
    fun `quit the last conversation`() {
        val s = getSession()
        val conversation = s?.conversation?.blockingList()?.firstOrNull()
        conversation?.blockingQuit()
        assert(s?.conversation?.blockingList()?.firstOrNull()?.id != conversation?.id)
    }


}